File: cpp/string/aho-corasick.cpp
#define NN 30
 
struct TrieNode {
  TrieNode(int size) : sz(size), cnt(0) {
    for(int i = 0; i < NN; ++i) next[i] = nullptr;
    suffix = nullptr;
  }
  TrieNode *next[NN];
  TrieNode *suffix;
  int sz, cnt;
};
 
TrieNode *root = new TrieNode(0);
 
void add(string s) {
  TrieNode *node = root;
  for(int i = 0; i < s.length(); ++i) {
    int code = s[i] - 'a';
    if(!node->next[code]) {
      node = node->next[code] = new TrieNode(node->sz + 1);
    } else {
      node = node->next[code];
    }
  }
  node->cnt = 1;
}
 
void suffix() {
  queue<TrieNode*> que;
  que.push(root);
  while(!que.empty()) {
    TrieNode* v = que.front(); que.pop();
    for(int i = 0; i< NN; ++i) {
      if(v->next[i] && v->sz < v->next[i]->sz) {
        if(v->suffix) {
          v->next[i]->suffix = v->suffix->next[i];
        } else {
          v->next[i]->suffix = root;
        }
        que.push(v->next[i]);
      } else {
        if(v->suffix) {
          v->next[i] = v->suffix->next[i];
        } else {
          if(root->next[i]) {
            v->next[i] = root->next[i];
          } else {
            v->next[i] = root;
          }
        }
      }
    }
  }
  root->suffix = root;
}

File: python/string/aho-corasick.py
# この実装のnodeの説明
#
# node[0] ~ node[25]:
#     'a' ~ 'z'が来た時の次のnode (failure関数の遷移先も含まれる)
# node[-3] (node[27]):
#     いずれかの文字の終端であるかを示す(1 = 終端である)
# node[-2] (node[28]):
#     このnodeが何文字目に位置するかを表す
# node[-1] (node[29]):
#     このnodeでfailureした時の遷移先

from collections import deque
base = ord('a')
root = [None]*30
root[-2] = 0; root[-3] = 0

# 文字列sの追加
def add(s):
    node = root
    for c in s:
        code = ord(c) - base
        child = node[code]
        if not child:
            node[code] = child = [None]*30
            child[-2] = node[-2] + 1
            child[-3] = 0
        node = child
    node[-3] = 1

# BFSでfailure関数のリンクを構築
def suffix():
    que = deque([root])
    while que:
        v = que.popleft()
        for i in range(26):
            if not v[i]:
                if v[-1]:
                    v[i] = v[-1][i]
                else:
                    v[i] = root[i] or root
                continue
            if v[-1]:
                v[i][-1] = v[-1][i]
            else:
                v[i][-1] = root
            que.append(v[i])
    root[-1] = root
